storage/src/test/scala/coop/rchain/storage/regex/PathRegexUnitTests.scala

```diff
@@ -0,0 +1,87 @@
+package main.scala.coop.rchain.storage.regex
+
+import org.scalatest._
+
+class PathRegexUnitTests extends FlatSpec with Matchers {
+
+  def doCheck(path: String,
+              tokens: Option[List[PathToken]],
+              //matchCases : Seq[(String, Seq[String])],
+              //compileCases : Seq[String],
+              options: PathRegexOptions = PathRegexOptions.default): Unit = {
+    val parsed = PathRegex.parse(path, options)
+    tokens.foreach(t => assert(parsed == t))
+  }
+
+  def check(path: String,
+            tokens: List[Any],
```

Comment:
We can't :( A lot of tests in tests in next part use mix of PathTokens and strings in a single list. I tried Either before using List[Any], and found that it is not a good choice as well, since even simple test full of "Left()"/"Right()" becomes completely unreadable. Unfortunately scala type system was created before TypeScript Type1|Type2 types :( I've made a quick search for implicit conversion from type1/type2 to Either[Type1, Type2] and found nothing readable enough to use in unit-tests.

Reply:
> A lot of tests in tests in next part use mix of PathTokens and strings in a single list

Oh that's unfortunate.  My hope is that we can have a more type-safe solution to this problem.  

I agree that `Either` is often a poor choice for this kind of situation.  

What about defining something like:

```scala
sealed trait Token
final case class PathToken(<existing members here>) extends Token
final case class StringToken(value: String) extends Token
```

Then we can use `List[Token]` instead of `List[Any]`.  This is probably the simplest solution.

Other, slightly more complicated, possibilities:
* shapeless' [`HList`](https://github.com/milessabin/shapeless/wiki/Feature-overview:-shapeless-2.0.0#heterogenous-lists)
* shapeless' [`Coproduct`](https://github.com/milessabin/shapeless/wiki/Feature-overview:-shapeless-2.0.0#coproducts-and-discriminated-unions) (these are closer to the TypeScript Union types you describe)